File: examples/basic/services/app-ts/src/utils/storage/key-verifier.ts
import { IVerifier, IHasher } from 'better-auth-ts'
import { Redis } from 'ioredis'
import { Secp256r1Verifier } from '../crypto/secp256r1.js'
import { Blake3Hasher } from '../crypto/blake3.js'
import { getSubJson } from './utils.js'

const HSM_IDENTITY = 'BETTER_AUTH_HSM_IDENTITY_PLACEHOLDER'

// needs to be server lifetime + access lifetime. consider a server that just rolled before the hsm
// rotation. it may issue a token 11:59:59 into the new hsm key's existence, but it's authorized with
// the old hsm key. that key is then valid for the access lifetime (15 minutes in our case) before
// it must be refreshed. again for app servers, this time must be server lifetime + access lifetime.
// for auth servers it is different. in this example, the access lifetime is a parameter of the store.
const TWELVE_HOURS_MS = 12 * 60 * 60 * 1000

interface LogEntry {
  id: string
  prefix: string
  previous: string | null
  sequenceNumber: number
  createdAt: string
  taintPrevious: boolean | null
  purpose: string
  publicKey: string
  rotationHash: string
}

interface SignedLogEntry {
  payload: LogEntry
  signature: string
}

export class KeyVerifier {
  private readonly client: Redis
  private readonly verifier: IVerifier
  private readonly hasher: IHasher
  private readonly cache: Map<string, LogEntry>
  private readonly accessLifetime: number

  constructor(redisHost: string, redisDbHsmKeys: number, accessLifetimeInMinutes: number) {
    this.accessLifetime = accessLifetimeInMinutes * 60 * 1000
    this.client = new Redis({
      host: redisHost.split(':')[0],
      port: parseInt(redisHost.split(':')[1]) || 6379,
      db: redisDbHsmKeys,
    })
    this.verifier = new Secp256r1Verifier()
    this.hasher = new Blake3Hasher()
    this.cache = new Map()
  }

  async verify(
    signature: string,
    hsmIdentity: string,
    hsmGenerationId: string,
    message: string
  ): Promise<void> {
    let cachedEntry = this.cache.get(hsmGenerationId)

    if (!cachedEntry) {
      // clear the cache, we are seeing a new key
      this.cache.clear()

      // Fetch all HSM keys from Redis
      const keys = await this.client.keys('*')
      const values = await this.client.mget(...keys)

      // Group by prefix
      const byPrefix: Map<string, Array<[SignedLogEntry, string]>> = new Map()

      for (const value of values) {
        if (!value) continue

        const payloadJson = getSubJson(value, 'payload')
        const record = JSON.parse(value) as SignedLogEntry
        const prefix = record.payload.prefix

        if (!byPrefix.has(prefix)) {
          byPrefix.set(prefix, [])
        }

        byPrefix.get(prefix)!.push([record, payloadJson])
      }

      // Sort by sequence number
      for (const [prefix, records] of byPrefix.entries()) {
        records.sort((a, b) => a[0].payload.sequenceNumber - b[0].payload.sequenceNumber)
        byPrefix.set(prefix, records)
      }

      // Verify data & signatures for all records
      for (const prefix of byPrefix.keys()) {
        const records = byPrefix.get(prefix)!
        for (const [record, payloadJson] of records) {
          const payload = record.payload

          if (payload.sequenceNumber === 0) {
            await this.verifyPrefixAndData(payloadJson, payload)
          } else {
            await this.verifyAddressAndData(payloadJson, payload)
          }

          // Verify signature over payload
          await this.verifier.verify(payloadJson, record.signature, payload.publicKey)
        }
      }

      // Verify chains
      for (const records of byPrefix.values()) {
        let lastId = ''
        let lastRotationHash = ''

        for (let i = 0; i < records.length; i++) {
          const payload = records[i][0].payload

          if (payload.sequenceNumber !== i) {
            throw new Error('bad sequence number')
          }

          if (payload.sequenceNumber !== 0) {
            if (lastId !== payload.previous) {
              throw new Error('broken chain')
            }

            const hash = await this.hasher.sum(payload.publicKey)

            if (hash !== lastRotationHash) {
              throw new Error('bad commitment')
            }
          }

          lastId = payload.id
          lastRotationHash = payload.rotationHash
        }
      }

      // Verify prefix exists
      const records = byPrefix.get(HSM_IDENTITY)
      if (!records) {
        throw new Error('hsm identity not found')
      }

      var tainted = false
      // Cache entries within 12-hour window (iterate backwards)
      for (let i = records.length - 1; i >= 0; i--) {
        const payload = records[i][0].payload

        if (!tainted) {
          this.cache.set(payload.id, payload)
        }

        if (payload.taintPrevious === true) {
          tainted = true
        } else {
          tainted = false
        }

        const createdAt = new Date(payload.createdAt)
        if (createdAt.getTime() + TWELVE_HOURS_MS + this.accessLifetime < Date.now()) {
          break
        }
      }

      cachedEntry = this.cache.get(hsmGenerationId)
      if (!cachedEntry) {
        throw new Error("can't find valid public key")
      }
    }

    if (cachedEntry.prefix !== hsmIdentity) {
      throw new Error('incorrect identity (expected hsm.identity == prefix)')
    }

    if (cachedEntry.purpose !== 'key-authorization') {
      throw new Error('incorrect purpose (expected key-authorization)')
    }

    // Verify message signature
    await this.verifier.verify(message, signature, cachedEntry.publicKey)
  }

  private async verifyPrefixAndData(payloadJson: string, payload: LogEntry): Promise<void> {
    if (payload.id !== payload.prefix) {
      throw new Error('prefix must equal id for sequence 0')
    }

    await this.verifyAddressAndData(payloadJson, payload)
  }

  private async verifyAddressAndData(payloadJson: string, payload: LogEntry): Promise<void> {
    // Serialize payload and replace id with placeholder
    const modifiedPayload = payloadJson.replace(
      new RegExp(payload.id, 'g'),
      '############################################'
    )

    const hash = await this.hasher.sum(modifiedPayload)

    if (hash !== payload.id) {
      throw new Error('id does not match hash of payload')
    }
  }

  async close(): Promise<void> {
    await this.client.quit()
  }
}

File: examples/basic/services/app-rb/lib/storage/key_verifier.rb
require 'redis'
require 'json'
require 'time'
require_relative '../crypto/secp256r1'
require_relative '../crypto/blake3'
require_relative './utils'

module Storage
  HSM_IDENTITY = 'BETTER_AUTH_HSM_IDENTITY_PLACEHOLDER'

  class LogEntry
    attr_reader :id, :prefix, :previous, :sequence_number, :created_at, :taint_previous, :purpose, :public_key, :rotation_hash

    def initialize(data)
      @id = data['id']
      @prefix = data['prefix']
      @previous = data['previous']
      @sequence_number = data['sequenceNumber']
      @created_at = Time.parse(data['createdAt'])
      @taint_previous = data['taintPrevious']
      @purpose = data['purpose']
      @public_key = data['publicKey']
      @rotation_hash = data['rotationHash']
    end
  end

  class SignedLogEntry
    attr_reader :payload, :signature

    def initialize(data)
      @payload = LogEntry.new(data['payload'])
      @signature = data['signature']
    end
  end

  class KeyVerifier
    def initialize(redis_host, redis_db_hsm_keys, server_lifetime_hours, access_lifetime_minutes)
      @redis = Redis.new(url: "redis://#{redis_host}/#{redis_db_hsm_keys}")
      @verifier = Crypto::Secp256r1Verifier.new
      @hasher = Crypto::Blake3.new
      @cache = {}
      @verification_window = server_lifetime_hours * 3600 + access_lifetime_minutes * 60
    end

    def verify(signature, hsm_identity, hsm_generation_id, message)
      cached_entry = @cache[hsm_generation_id]

      unless cached_entry
        # Clear cache before repopulating
        @cache.clear

        # Fetch all HSM keys from Redis
        keys = @redis.keys('*')
        raise 'No HSM keys found in Redis' if keys.empty?

        values = @redis.mget(*keys)

        # Group by prefix
        by_prefix = {}
        values.compact.each do |value|
          payload_json = get_sub_json(value, 'payload')
          record = SignedLogEntry.new(JSON.parse(value))
          prefix = record.payload.prefix

          by_prefix[prefix] ||= []
          by_prefix[prefix] << [record, payload_json]
        end

        # Sort by sequence number
        by_prefix.each do |prefix, records|
          by_prefix[prefix] = records.sort_by { |r, _| r.payload.sequence_number }
        end

        # Verify data & signatures for all records
        by_prefix.each do |prefix, records|
          records.each do |record, payload_json|
            payload = record.payload

            if payload.sequence_number.zero?
              verify_prefix_and_data(payload_json, payload)
            else
              verify_address_and_data(payload_json, payload)
            end

            # Verify signature over payload
            @verifier.verify(record.signature, payload.public_key, payload_json.bytes)
          end
        end

        # Verify chains
        by_prefix.each_value do |records|
          last_id = ''
          last_rotation_hash = ''

          records.each_with_index do |(record, _), i|
            payload = record.payload

            raise 'bad sequence number' if payload.sequence_number != i

            unless payload.sequence_number.zero?
              raise 'broken chain' if last_id != payload.previous

              hash = @hasher.sum(payload.public_key.bytes)

              raise 'bad commitment' if hash != last_rotation_hash
            end

            last_id = payload.id
            last_rotation_hash = payload.rotation_hash
          end
        end

        # Verify prefix exists
        raise 'hsm identity not found' unless by_prefix.key?(HSM_IDENTITY)

        records = by_prefix[HSM_IDENTITY]

        # Cache entries within 12-hour window (iterate backwards)
        tainted = false
        records.reverse_each do |record, _|
          payload = record.payload
          @cache[payload.id] = payload unless tainted

          tainted = payload.taint_previous || false

          break if payload.created_at + @verification_window < Time.now
        end

        cached_entry = @cache[hsm_generation_id]
        raise "can't find valid public key" unless cached_entry
      end

      raise 'incorrect identity (expected hsm.identity == prefix)' if cached_entry.prefix != hsm_identity
      raise 'incorrect purpose (expected key-authorization)' if cached_entry.purpose != 'key-authorization'

      # Verify message signature
      @verifier.verify(signature, cached_entry.public_key, message.bytes)
    end

    private

    def verify_prefix_and_data(payload_json, payload)
      raise 'prefix must equal id for sequence 0' if payload.id != payload.prefix

      verify_address_and_data(payload_json, payload)
    end

    def verify_address_and_data(payload_json, payload)
      # Serialize payload and replace id with placeholder
      modified_payload = payload_json.gsub(payload.id, '############################################')

      hash = @hasher.sum(modified_payload.bytes)

      raise 'id does not match hash of payload' if hash != payload.id
    end

    def close
      @redis.quit
    end
  end
end
